### ISSUE TITLE
Establishing expectations for registered packages

### ISSUE THREAD

**Issue by ararslan** [JuliaLang/METADATA.jl#6761]
There appears to be some dissatisfaction with the speed—be it too fast or too slow—of the METADATA workflow. I think this may stem from a lack of a common understanding of what's expected for registered packages. To that end, I think it may be a good idea to have a discussion about what exactly is expected of registered packages, including what is required for registration, so that everyone is on the same page. I think having a standard pull request template, as suggested in #5307, may help with this, but it would be nice to have consensus on expectations in the interim.

Note that this is **not** intended to be a debate on the merits of METADATA or registration, nor is it a suggestion that we should disallow certain packages or kinds of packages. I'm merely looking for input on what is/should be expected and how everyone can come to a mutual understanding to avoid future drama and/or confusion.

**Comment by tkelman** [comment]
Standard things I check for when reviewing a new package:

- the guidelines listed here https://github.com/JuliaLang/METADATA.jl/blob/metadata-v2/README.md and here http://docs.julialang.org/en/release-0.4/manual/packages/#guidelines-for-naming-a-package
- do the git tags match the metadata tags (the tests added in https://github.com/JuliaLang/METADATA.jl/pull/6485)
- if the package can easily be made to pass its tests on Travis and/or AppVeyor, are they enabled and passing for all supported Julia versions?
- is any code clearly derived from or copied from some other project without appropriate license attribution?
- is the package doing anything obviously malicious, especially in a `deps/build.jl` script
- is every dependency that is used via `using`, `import`, or `importall` listed in `REQUIRE` or `test/REQUIRE`, preferably with minimum version requirements?
- is the package changing the behavior of other unrelated code via "type piracy" - aka extending methods from Base or some other package on type signatures that are entirely from Base or some other package. If either the method or at least one of the input types is owned by the same package, then there's no problem.
- is the package doing any fragile simulation of conditional dependencies that will interact poorly with precompilation - if so, it should explicitly mark itself as `__precompile__(false)`

Other far less important nitpicky things that I'll comment on but can easily be fixed in future tags:
- using `Pkg.dir` to refer to a package's own location instead of `dirname(@__FILE__)`
- not being specific enough about `VERSION` conditions
- `Pkg.test` making the package dirty or not cleaning up temporary test files

These were off the top of my head, I'll add more if I can think of them. If I've nitpicked about your package in a way that didn't fit in one of the above categories, feel free to refresh my memory.

I would very much like to see Pkg3 implement namespaces (e.g. #2417) so people can do whatever they want in their own personal free-for-all unofficial registries that can be caveat emptor opted into, and have the official enabled-by-default package registry contain curated packages that have been vetted for stability via the above criteria (e.g. #2777 and #2800).

**Comment by tbreloff** [comment]
First off: Tony I thank you for taking on a thankless and herculean effort to have total stability in the package landscape.  Having a curated source for package acquisition is key to the future of Julia.

The problem, as I see it, is just that there's too much focus on METADATA as the only source for packages.  Julia package development is moving at way too fast a speed.   We've been talking about Pkg3, and different levels of curation, etc for a long time, but it's not totally clear what that will eventually look like.

For anyone interested in a non-Pkg3 alternative to managing packages, please come discuss at https://gitter.im/tbreloff/MetaPkg.jl and https://github.com/tbreloff/MetaPkg.jl/issues.  I would like to collaborate with anyone that shares my need for very fast development cycles involving many inter-related packages.  **Pkg/METADATA should stay exactly as is**, and remain the "super-robust and curated" source for packages.  Tony: keep up the good work.  Your "standard" and "nitpicky" lists are exactly what we need.  I just want an easy alternative for "releasing" in between METADATA tags.

**Comment by tkelman** [comment]
What does "releasing" mean to you though? I really strongly discourage you from having your users checking out master of a bunch of different packages, unless they're actively contributing PR's or wanting to test unstable code. Witness what the Go community struggled with due to lack of release versioning for so long, it was terrible for reproducibility of anything.

**Comment by tbreloff** [comment]
I explained a lot in the MetaPkg issues.  I'm happy to continue the
discussion over there, and reserve this issue for METADATA-specific
discussion. Certainly I want reproducible builds. But I also want easy
certification of true compatibility between packages, which I don't think
we've achieved yet.

**Comment by tkelman** [comment]
Best way to do that is tag packages one at a time so Travis can actually test with the right dependency versions. If a tag breaks dependent packages, then it should be adding upper bounds to past tags of those dependents.

**Comment by tkelman** [comment]
> I explained a lot in the MetaPkg issues

I don't see anything there that answers the question. What is your reproducible "release" mechanism that metadata doesn't cover? I very much disagree that Pkg and metadata should stay as is, this isn't scalable and they both need to be redesigned to meet more needs, including yours if you could articulate them in a better way than issues on your own experimental repo. It sounds like you mostly want multiple stability levels of release channels, which is something that namespaces should be well suited to handle.

**Comment by tbreloff** [comment]
I'll quickly summarize a few ideas:

- Package "bundles" should be first-class... groups of packages that are inter-related could be tagged and tested as a group, and the tag can apply to all of the packages.  Not only does this reduce the churn of individual tags, but it gives an extra level of robustness for quirky breaking changes.
- It should be possible to easily depend on unregistered packages, forks of registered packages, or anything else... maybe that dependency will preclude a "curated seal of approval", but we shouldn't make this hard on purpose.
- QA should be spread out to github orgs, instead of one central place.  Let each org control their package ecosystem as they see fit.  Reliability will depend on the org (and users will quickly come to know which orgs are reliable), but I think that if we put trust in the organizations and give them the right tools, that they'll be able to actively manage dependency and QA issues better than an outside party.
- There should be a tiered-METADATA of some kind, with different requirements for inclusion in each tier.  It should be possible to register whole bundles with each tier, and the maintainers of those curated lists would be:
  - a double-check on the QA process
  - a certification that one bundle doesn't sneakily break another bundle
  - a docker-style snapshot tagger, which will give reproducible environments for super-fast setup of a known-working set of packages.

The key things that we have to solve:

- Make sure packages play well together
- Make sure the QA process isn't so onerous that no one partakes
- Give users fast ways to get set up in robust environments
- Make it easy to manage organizations/packages that are in active development

I don't quite see how "namespaces" alone will solve all of these issues.  However I'm positive that I'm not in the inner circle of what Pkg3 and alternatives will look like, so it's hard for me to know what the plans are.  If there are public discussions about that, please link to them!  As I said, I'll be experimenting with some of these concepts in MetaPkg.  If they turn out to be good, I hope they help shape the development of Pkg3, if not then no big deal.  Either way, I'd prefer to move this discussion somewhere else so as not to spam this repo.

**Comment by ChrisRackauckas** [comment]
I think Tom highlights two things that should really be taken into account:

- The needs of developers can be different than the needs of users. This fact is still true in Julia even though the lines between "developer" and "user" is heavily blurred. The current METADATA setup is good for users, but can be rough on developers. Clearly defined "bundles" make it much easier to find out how to develop in a certain package sphere. For example, the first step to finding out how to get on "the development branch of Juno" is finding out all of the packages that need to be currently checked out to make it work. Since the packages are moving so rapidly, It would be nice if there was a more central way to handle agile workflows where many different things are changing at once. And again, since there isn't as much separation between developer and user, being able to easily swap hats is necessary.

- Organization-based management. Orgs have done an unusually good job of helping police the package system. Because of this, I think that the orgs should be more central to Julia's package system. While not a complete "standard library", org-based metapackages, `using Math`, `using Plots` (well, that's basically an org), `using DiffEq`, can give pretty complete functionality, are developed in unison, and made to play well with each other.

While you have done a very good job @tkelman and I have learned a lot from you, as you say the current system isn't scaling too well and I really think some way of using orgs to your advantage could help.

**Comment by KristofferC** [comment]
This seems like a good place to have the discussion? Or the main Julia repo I guess.

**Comment by tkelman** [comment]
I don't know if there are plans for package bundles as explicit entities, but unregistered dependencies (which would prevent inclusion in the curated metadata, but be allowed elsewhere) and tiered metadata should be covered by Pkg3. A formal design proposal is extremely overdue there, if that continues to be the case then we'll propose something different to accomplish the same goals sooner.

If you frequently need to check out master branches of things, that's a sign that you aren't tagging often enough. Mark new things as experimental, but inventing a parallel package management scheme that encourages more people to be on master of more packages doesn't seem like a good solution here.

Tagging a set of packages all simultaneously makes testing and QA harder, not easier. Package dependencies are managed and versioned individually. Dependency version bounds are set on a per package level. If someone applies bounds on a single member of the group of packages you tagged all at the same time, then the resolver will give them a different set. If you the package author are not careful about your bounds between your packages, then users can get feasible version sets that won't work together.

Having users and developers using different sets of tools also means the testing you do as a developer won't be representative of what your users are installing. Imposing artificial barriers here and widely different workflows for authors vs users is not something I consider healthy or desirable.

**Comment by ChrisRackauckas** [comment]
>Mark new things as experimental, but inventing a parallel package management scheme that encourages more people to be on master of more packages doesn't seem like a good solution here.

This is where conditional dependencies come in. It's rough to have something experimental and "part of the main package" because if that little tiny part of the package has a problem (and it's changing rapidly), then the entire package cannot be used. If it's on master branch and things change around, you can throw in a quick fix (and most users will not be using that version). With conditional dependencies, it would isolate the problem to be only when that feature is used.

**Comment by tkelman** [comment]
Then put the experimental part somewhere else? Its own package or sub module? Make it throw an error unless the user sets a module level boolean flag?

I get that you are convinced you really want them, but conditional dependencies are just bad, they're bad in C and autotools and they're going to be a pain even when there's an implementation that works correctly in julia. Compiling a half-functional version of something because the user was missing various optional dependencies is usually worse than being explicit about whether you want just the bare minimum version or the full version.

**Comment by tbreloff** [comment]
> If you frequently need to check out master branches of things, that's a sign that you aren't tagging often enough.

Of course, but it's simply not possible to tag as often as I'd like when the process is as manual and delayed as it is now.   For Plots, whenever I push to master, it is effectively a "tag" in my mind.  I work on the dev branch, and I have a short script which merges into master and pushes.  It takes 3 seconds.  To be frank, I don't have the time to spend an hour getting tags for Plots every day or 2, and many users want to immediately use the features that I develop.  tl;dr Tagging needs to be near-zero effort if you want frequent tagging.

> Tagging a set of packages all simultaneously makes testing and QA harder, not easier. Package dependencies are managed and versioned individually. Dependency version bounds are set on a per package level. If someone applies bounds on a single member of the group of packages you tagged all at the same time, then the resolver will give them a different set. If you the package author are not careful about your bounds between your packages, then users can get feasible version sets that won't work together.

All of this is a sign that the METADATA model is fragile and unscalable.  If "bundles" were first class, and package authors set bounds on bundles instead of individual packages, I think there's less room for undefined package states.

> Having users and developers using different sets of tools also means the testing you do as a developer won't be representative of what your users are installing. Imposing artificial barriers here and widely different workflows for authors vs users is not something I consider healthy or desirable.

I completely agree.  Another reason why its so painful now when people want the latest features, and their only option is to manually check out a bunch of packages that they know nothing about, with no easy way to revert to a clean install (manually freeing packages is tedious, error-prone, and frequently forgotten).  Again, if bundles were first class, you could checkout/free a bunch of related packages with one command, and I think users would more quickly revert to a tagged release.

> conditional dependencies are just bad

This could be a long discussion, but there are many projects/orgs out there that need some sort of conditionality. (Plots, JuMP, Images, DataFrames, DifferentialEquations, and JuliaDB are a few packages that come to mind.)  Frequently there is a core package/api, and a bunch of optional add-ons and backends for calculations, graphics, or whatever else.  Frequently, those optional dependencies are both heavyweight and sparsely used (so it doesn't make sense to automatically install them).  You frequently make the suggestion to pull the optional code into its own package, but this is both time-consuming and onerous for developers and sub-optimal for users.  The point is that we need a way to separate optional functionality (with optional dependencies) within a larger framework, but still tag and release that larger framework as a whole.  Tagged bundles would greatly help the landscape here.

**Comment by tkelman** [comment]
Tagging is 2 files and a git tag, if you're staying within the above guidelines. Your workflow of having separate dev vs master branches is very uncommon among package authors and I don't think it should drive complexity in the package manager when there are already mechanisms that work, if you would use them. If people want to use the latest features, then the author should tag them. It'll be far less error prone that way than inventing a parallel distribution scheme.

Package version resolution isn't part of metadata's design, it's how any modern package manager works. A handful of them may have package bundles (build-essential etc) but usually implemented within the existing mechanisms. You can go ahead and make an empty package that pins a set of dependencies in its require file today, and there you have your first class package bundles. What are you referring to as "undefined states"? There are version combinations allowed by bounds and there are version combinations that are infeasible because of bounds.

If you want things to always be bundled and move in lockstep, why are they separate packages? You want to split functionality into a bunch of grouped packages, but you don't want to move optional components into their own packages?

The way JuMP does things is both solvers and modeling environments code to the same intermediate MathProgBase API, then they don't need to depend on each other at all. The user requests exactly which solver they want, and it all works. Other organizations should be learning from it, not reimplementing dozens of fragile ad hoc approximations. I'll also note that JuliaOpt has also been correctly applying upper bounds, so packages can all be tagged individually without breaking anything. Code for totally separate optional backends shouldn't all live in the same place.

**Comment by timholy** [comment]
Just to chime in and say that both viewpoints have merit. Tagging may seem simple, but as someone who does it a lot I can assure you that somehow it does still take nontrivial time and effort, especially when mistakes need to be corrected. It's something I'm willing to do, but that 5-15 minutes sure would have been nice to spend in some other way. Note that the majority of core julia devs who own packages rarely tag them; if that group can't be persuaded to do it, that's a clear sign that the workflow isn't perfect (and that beating on people won't fix it).

For me personally, the more annoying aspect with our current system is that it can lead to situations in which CI in one package requires waiting for a tag another package. I work with functionality that is spread across many different packages, and I tend to think it's good to split functionality into relatively "atomic" pieces, but it does mean that you can get into pileups where a 10-line change spread across 3 packages can take two days (or more) to make its way through the pipeline. To be frank, that's completely unacceptable. I usually cheat somehow: edit `.travis.yml` to check out the master branch and get them all through CI in one fell swoop (my least favorite option, because I'm always afraid that I'll forget to change it back) or use my METADATA superpowers to do PR/tag/PR/tag/PR/tag quickly (esp. if I'm working early in the morning when the Travis queue is short and the METADATA PRs complete CI quickly). It's hard to overstate what a massive disincentive this is to julia development: there are plenty of studies that show that people generally don't respond well when the "real work" component of a task is simple (it's my repo, and I can push commits that fix a genuine problem) but the administrative overhead is painful. (It's also well-established that people tend to underestimate the amount of communication needed when you're working in the context of a large organization, so I'm not trying to say the "fault" is entirely one-directional.)

Moreover, while I agree that version resolution is important in principle, I don't think it's safe yet to make as extensive use of that feature as might be attractive in principle: until we have tools that walk users through an explanation of exactly which packages are conflicting and how to select versions that resolve the issues, I actually think that having too many constraints is likely to tie users Pkg installations up in knots. That said, I also agree that having version resolution is a crucial long term requirement.

I greatly appreciate the unbelievable skill and attention that @tkelman puts in to this, while also finding this whole system to be one of the more annoying aspects of julia development. To be honest, I'd almost favor a free-for-all on METADATA if it gave @tkelman some developer time to help make the whole system work better. We need:

- automatic detection of "cross-ecosystem" tagging needs given the current state of a developer's `.julia/v0.5` folder
- something like a "staged tags" system so that CI can solve the problem of synchronized changes across packages
- no human effort required to check whether a given METADATA change is correct: if it passes travis CI, it should merge immediately (and automatically) (new package registrations can be an exception, of course)
- ideally those checks should be easy to run on the developer's machine
- sophisticated & helpful resolution of version conflicts for users

**Comment by tkelman** [comment]
> if it passes travis CI, it should merge immediately

I disagree with this, given the possibility for malicious code. It isn't something that's happened yet, but given `deps/build.jl` scripts execute nearly arbitrary code on users' machines when they run `Pkg.update`, the default metadata should really have a second set of eyes quickly checking for anything out of the ordinary.

**Comment by tkelman** [comment]
> especially when mistakes need to be corrected

I'll gladly review any proposed implementation of https://github.com/JuliaLang/PkgDev.jl/issues/39. Likewise, we can make `Pkg.status()` show more information about version bounds, the information's all there but not much of it is being shown currently.

A free-for-all sounds great in theory until the whole thing breaks horribly with a left-pad style situation, or worse corruption. I don't look forward to fixing that kind of mess after it's been made. Free-for-alls that are opt-in and easily reversible to get yourself out of when things break, and primarily used as a way of getting brand new / experimental stuff, I'm all for. Pkg3 needs to enable this, and I'm sure I'll be helping implement and test whichever parts of the eventual proposal that I happen to agree with. If waiting for a proposal is proving an ineffective way of changing things, we can propose less ambitious incremental steps. More thorough automated testing, more features in PkgDev for provisionally testing not-yet-tagged packages (hint - if you're doing anything temporary on travis, make a new short-lived branch for it), whatever.

**Comment by tbreloff** [comment]
Thanks @timholy for chiming in.  I agree with everything you said (of course, I usually do)

> For me personally, the more annoying aspect with our current system is that it can lead to situations in which CI in one package requires waiting for a tag another package. I work with functionality that is spread across many different packages, and I tend to think it's good to split functionality into relatively "atomic" pieces, but it does mean that you can get into pileups where a 10-line change spread across 3 packages can take two days (or more) to make its way through the pipeline. To be frank, that's completely unacceptable.

To be clear, this is exactly what annoyed me so much with the last PR.  I had a fully functional tag that improved what was out there, but there was a chicken-and-egg problem of requiring the PR to be merged before the travis tests would pass.  Now I could have closed that PR, retagged PlotThemes with the 4-character change to one of the tests, manually push the tags, branch METADATA, push to my fork, go to the website, submit the PR, wait a day for it to be merged, then rerun the Plots tests, wait 4 hours for the travis tests to finish, then tag, push the tags, branch METADATA, push to my fork, submit a new PR, and wait another day for it to be merged.  Much of the frustration stems from how manual the process is... if it was better automated and much of the testing/verification could happen before the PR is submitted, then it wouldn't be so painful.

> > if it passes travis CI, it should merge immediately

> I disagree with this, given the possibility for malicious code.

I agree with @tkelman here... merges shouldn't happen automatically.  But there should be very little human involvement beyond verifying that there's not a malicious attack that is tricking the auto-verification process.

> The way JuMP does things is both solvers and modeling environments code to the same intermediate MathProgBase API, then they don't need to depend on each other at all. The user requests exactly which solver they want, and it all works. Other organizations should be learning from it, not reimplementing dozens of fragile ad hoc approximations

Just so you know, JuMP has a similar approach to its conditional dependencies as Plots.  Basically, there is no dependency... dynamically there is an `@eval import <solver>`, and if that fails then there's a warning that it might not be installed properly.  JuMP's design could be improved greatly and made precompile-friendly with proper conditional modules.  Currently you have to choose between "eval hacks" or a frustrating user experience.  JuMP and Plots have chosen eval hacks.

**Comment by tkelman** [comment]
There's only an eval import in the tests and in the MathProgBase default setting, which is likely going to be removed - https://github.com/JuliaOpt/MathProgBase.jl/issues/121. Better to specify the optional component that you actually want to use.

**Comment by tbreloff** [comment]
> Better to specify the optional component that you actually want to use

I'm trying to make sense of what you think is best practice at the moment for a JuMP-style framework.  Let me know if you agree that this is what you're suggesting:

- A generic interface to be implemented by specific backends (MathProgBase)
- A high-level package which depends on and uses that generic interface (JuMP)
- A bunch of distinct and independent "backend wrappers" which depend on the generic interface and anything specific to that backend (SCS, etc)

Then a user would have to do:

```julia
using JuMP, SCS
solver = SCSSolver(...)
solve(..., solver=solver)
```

and there would be no option for a default solver anymore.

So... I think this is mostly fine, and the worsened user-experience can be helped using global variables that track a "current" solver.  In fact, I would follow this pattern with Plots if it wasn't for METADATA, but I still may refactor.  Like I've been saying, if there were nice ways to quickly tag and publish all of JuliaPlots at once, then I'd be more willing to split into 10 extra packages to solve the lack of optional dependencies and conditional modules.

**Comment by tkelman** [comment]
You set the solver when you construct a model object in JuMP, but yeah that basic pattern.

PkgDev was separated from Base so it can be changed independently. If you have suggestions for tagging package sets and you want them to be widely used, that's where I'd focus on making improvements. As far as testing goes, we could add an argument to `PkgDev.publish` and/or `PkgDev.tag` that pushes a modified `.travis.yml` to a test branch where `Pkg.init()` runs from the same branch of your METADATA fork as the PR, and add links to the branch statuses to the autogenerated commit message.

**Comment by timholy** [comment]
>  given the possibility for malicious code

hmm, very important point. But then I think we need to rethink this a bit. Let me make an analogy: suppose you discover a security flaw in firefox that stems from a problem in OpenSSL. Should the OpenSSL/firefox developers have to wait for two new Debian point releases to test and distribute the fix?

There are so many flaws with this analogy it's not even funny, but it does illustrate an important point: the security needs of distributing code to users are very different from the efficiency needs of developers. Since Travis is sandboxed, we don't need to worry about infecting them, and since developers run a lot of untriaged code we are on our own anyway. For the benefit of both communities, we need to decouple the process of advancing the code base from the necessarily slower process of security review.

**Comment by tkelman** [comment]
I'll add an autogenerated comparison link to the PkgDev commit message so it's just one click (instead of the little bit of typing that I do) to quickly skim the changeset for a new tag. I don't think we should take the human entirely out of the loop of the primary Metadata, but we can do a few more things to minimize the manual work per tag. Debian doesn't distribute things that haven't been reviewed, neither do Homebrew or CRAN. Npm and PyPI do, but I believe those package managers don't operate in auto-upgrade mode by default.

**Comment by tbreloff** [comment]
> For the benefit of both communities, we need to decouple the process of advancing the code base from the necessarily slower process of security review.

And this gets back to the (IMO) dire need for a easy way to manage living (i.e. actively developed) packages/organizations.  We can't wait 6 months for Pkg3 (which hasn't been designed yet AFAIK, so there's no guarantee it will be a magic bullet).  It's not enough to just add a few extra methods to PkgDev because that still implies that METADATA is the one-true-source for packages.  I've got some tools in MetaPkg primarily to be able to provide conveniences around the very-manual Pkg commands that are currently required to checkout/clone/free groups of packages all at once.  **I don't want to create a completely separate package manager**  I just want package management to be uber-convenient like everything else I work on.

**Comment by tkelman** [comment]
> METADATA is the one-true-source for packages

Changing this requires rewriting the package manager. Pkg3 has been designed but that design has not yet been shared with anyone.

If you're writing convenience methods for cloning things that are outside of metadata, then either you're always on master (or other non-pinned branches) of those packages in which case that's not suitable for users to touch at all, or you're inventing an alternate mechanism of versioning in which case yes that's a completely separate package manager.

**Comment by tbreloff** [comment]
Tony, I have **tons** of repos that are not registered with METADATA.  Lots of them I would register if the process wasn't so time consuming.  Some are works in progress that are still usable (similar to Julia itself).  The blanket statement that users should never have to checkout a package seems a little naive to me.

Also:

> Pkg3 has been designed but that design has not yet been shared with anyone.

This really really bothers me.

**Comment by ChrisRackauckas** [comment]
>Pkg3 has been designed but that design has not yet been shared with anyone.

Why not have the design in an issue for others part read/participate to make sure it meets everyone's needs? I can see the reason for some to keep executive control (planning by committee just takes too long), but I don't see why it wouldn't be out there for others to be able to plan for.

>and there would be no option for a default solver anymore.

That's my big concern. I am jumping into a realm of choosing defaults from hundreds of algorithms, where the default takes into account things like the requested tolerances and properties of a user's input function. I don't see how a global constant could set a default like this. We are getting closer in DiffEq land though.

**Comment by ChrisRackauckas** [comment]
>To be clear, this is exactly what annoyed me so much with the last PR. I had a fully functional tag that improved what was out there, but there was a chicken-and-egg problem of requiring the PR to be merged before the travis tests would pass.

This is an even worse problem when you're waiting on other people to tag.

**Comment by ararslan** [comment]
> Some are works in progress that are still usable (similar to Julia itself). The blanket statement that users should never have to checkout a package seems a little naive to me.

FWIW I completely agree that users should never _have_ to check out a package. In my mind at least, having a package checked out to master means "I'm consciously opting in to using _potentially untested, experimental, beta-level features_ and I understand the risks." Once a package is tagged, even with a patch version, that sends a signal that the code as of a particular commit is then ready for mainstream use.

> Of course, but it's simply not possible to tag as often as I'd like when the process is as manual and delayed as it is now. For Plots, whenever I push to master, it is effectively a "tag" in my mind. I work on the dev branch, and I have a short script which merges into master and pushes.

I definitely understand the frustration with the work required to have things tagged. I think there's plenty of room for the process and machinery to be improved. I also think it's important to note though that many folks keep master in a "shippable" state, which is made easier by the use of continuous integration and thorough tests, but many don't. That kind of workflow, and even the presence of tests, depends entirely on the developer in question. I think that really underscores the importance of tags as a common, agreed-upon way of marking code as stable. In general I tend not to use code from the master branch unless I know _exactly_ what I'm getting; otherwise I use the latest tag.

**Comment by StefanKarpinski** [comment]
@tkelman wrote:

> Pkg3 has been designed but that design has not yet been shared with anyone.

This is a dangerous and alarming way to put this. Pkg3 does not have a final design – it's not going to leap from my skull, fully armored like Athena. I have a largely complete – but not final – sketch in my head based on discussions over the summer with @wildart and subsequent work on my own. I'm working very hard on getting into a coherent enough written form to publish. It's a sprawling, confusing thing to design and imposing order on it while addressing all the various concerns is not exactly easy. When it's ready (literally any day now), I will publish it and then everyone can discuss, debate, argue, etc. I'm sure that the final version based on community feedback will end up being quite different and much better than what I'm initially going to propose.

Perhaps I should have taken a more public, community-driven approach to this (after many sleepless nights working on this, I'm really starting to regret not doing so), but to be perfectly frank, I didn't largely because certain people have a tendency to nitpick any idea that's less than perfect to death. Since all ideas start out half-baked, I felt that that would get us nowhere fast.

**Comment by ChrisRackauckas** [comment]
Thanks @StefanKarpinski for the clarification. I can't wait to read it!

> but to be perfectly frank, I didn't largely because certain people have a tendency to nitpick any idea that's less than perfect to death.

Perfectly understandable. A lot of times big committees take time and usually hurt more than they help (this is why I stay away from Base). I think moreso than participating in any conversation, I'd like to just get a sense of what's going on. For some of these things like "standard library"/the parallelism changes, there's mentions of it here and there, but it's hard to tell whether there is something actually in the works or if it's a "ehh, we actually don't care about this, don't expect it anytime soon." Whatever the choice is, it's fine (you doing great work and everything is just a choice of what awesome thing to do with your time), but it helps to keep us posted on what to expect.

**Comment by StefanKarpinski** [comment]
At this point I can either spend time explaining what to expect or writing it up – the latter seems more productive, so I'm going to do that.

**Comment by tkelman** [comment]
I'm merely paraphrasing your own recent private status updates of being done with the design and in the process of writing it up. It's clear this is too important to not do anything about - the design will require iteration, but uncertainty waiting for a public starting point isn't doing much good.

**Comment by sbromberger** [comment]
From @timholy at https://github.com/JuliaLang/METADATA.jl/issues/6761#issuecomment-254172949:
> Tagging may seem simple, but as someone who does it a lot I can assure you that somehow it does still take nontrivial time and effort, especially when mistakes need to be corrected. It's something I'm willing to do, but that 5-15 minutes sure would have been nice to spend in some other way. 

Quoted for truth.

**Comment by StefanKarpinski** [comment]
Incomplete, but here's my writeup of Pkg3 so far: 

https://github.com/JuliaLang/Juleps/blob/master/Pkg3.md

This is neither complete, nor meant to be final, but rather as a starting point for feedback, discussion and design iteration.

**Comment by tkelman** [comment]
Where and in what form should that feedback happen, exactly?

**Comment by StefanKarpinski** [comment]
Issues and/or PRs?

**Comment by StefanKarpinski** [comment]
Interpretive dance is also acceptable.

**Comment by quinnj** [comment]
👯

**Comment by DilumAluthge** [comment]
METADATA.jl is no longer maintained, so I am closing this issue.

If this issue is still relevant, please re-open a new issue on the Julia General registry: https://github.com/JuliaRegistries/General
